### PR TITLE
lseek_syscall and close_syscall updates

### PR DESCRIPTION
**Pull request by namanlalitnyu**
## Description

Fixes # (issue)
The following changes include the tests and comments in the code for the `lseek_syscall` and `close_syscall` file system calls under RustPosix.
The tests were added to cover all the possible scenarios that might happen when calling the file system_calls  `lseek_syscall` and `close_syscall`.

### Type of change
- []  This change just contains the tests for an existing file system call.
- []  This change contains the minor code changes and comments for lseek_syscall and close_syscall.
- []  This change contains code reformatting for existing file system calls. 

## How Has This Been Tested?
Inorder to run the tests, we need to run `cargo test --lib` command inside the `safeposix-rust` directory.

All the tests are present under this directory: `lind_project/src/safeposix-rust/src/tests/fs_tests.rs`

- Test A - `ut_lind_fs_lseek_on_file()`
- Test B - `ut_lind_fs_lseek_on_directory()`
- Test C - `ut_lind_fs_lseek_invalid_whence()`
- Test D - `ut_lind_fs_lseek_beyond_file_size()`
- Test E -  `ut_lind_fs_lseek_before_start_of_file()`
- Test F - `ut_lind_fs_lseek_on_pipe()`
- Test G - `ut_lind_fs_lseek_on_chardev()`
- Test H - `ut_lind_fs_lseek_on_epoll()`
- Test I - `ut_lind_fs_close_chardev()`
- Test J - `ut_lind_fs_close_pipe()`
- Test K - `ut_lind_fs_close_socket()`
- Test L - `ut_lind_fs_close_directory()`
- Test M - `ut_lind_fs_close_regular_file()`

## Checklist:
- [] My code follows the style guidelines of this project
- [] I have commented my code, particularly in hard-to-understand areas
- [] My changes generate no new warnings
- [] I have added tests that prove my fix is effective or that my feature works
- [] Any dependent changes have been added to a pull request and/or merged in other modules (native-client, lind-glibc, lind-project)

### CODE REVIEW

File: src/safeposix/syscalls/fs_calls.rs [hunk 1]
```diff
@@ -2443,62 +2443,164 @@ impl Cage {
...
+        //Err(), unwrapping on which  produces a 'panic!'
+        //otherwise, file descriptor table entry is stored in 'checkedfd'
         let checkedfd = self.get_filedescriptor(fd).unwrap();
+        // Acquire a write lock on the file descriptor to ensure exclusive access.
         let mut unlocked_fd = checkedfd.write();
+
+        // Check if the file descriptor object is valid
         if let Some(filedesc_enum) = &mut *unlocked_fd {
             //confirm fd type is seekable
             match filedesc_enum {
```
Comment (davidge20) [inline src/safeposix/syscalls/fs_calls.rs:2506]: good explanations here for each file type

File: src/safeposix/syscalls/fs_calls.rs [hunk 2]
```diff
@@ -2443,62 +2443,164 @@ impl Cage {
...
                     let inodeobj = FS_METADATA
                         .inodetable
                         .get(&normalfile_filedesc_obj.inode)
                         .unwrap();
 
-                    //handle files/directories differently
+                    // Match the type of inode object with the type (File, Socket, CharDev, Dir)
+                    // Handle files/directories differently
                     match &*inodeobj {
                         Inode::File(normalfile_inode_obj) => {
```
Comment (davidge20) [inline src/safeposix/syscalls/fs_calls.rs:2546]: good descriptions here as well

File: src/safeposix/syscalls/fs_calls.rs [hunk 3]
```diff
@@ -2443,62 +2443,164 @@ impl Cage {
...
                                 return syscall_error(
                                     Errno::EINVAL,
                                     "lseek",
                                     "seek to before position 0 in file",
                                 );
                             }
-                            //subsequent writes to the end of the file must zero pad up until this
-                            // point if we overran the end of our file
-                            // when seeking
+                            // subsequent writes to the end of the file must zero pad up until this
```
Comment (davidge20) [inline src/safeposix/syscalls/fs_calls.rs]: What do you mean by zero pad up ?

File: src/safeposix/syscalls/fs_calls.rs [hunk 4]
```diff
@@ -3058,91 +3144,180 @@ impl Cage {
...
+            // We decide, how to proceed depending on the fd type.
+            // First we check in the file descriptor to handle stream / epoll (no-op),
+            // sockets (clean the socket), and pipes (clean the pipe), and if it is a
+            // normal file descriptor we decrement the reference count to reflect
             // one less reference to the file.
             match filedesc_enum {
                 //if we are a socket, we dont change disk metadata
-                Stream(_) => {}
-                Epoll(_) => {} //Epoll closing not implemented yet
+                Stream(_) => {} // Stream closing not supported
```
Comment (davidge20) [inline src/safeposix/syscalls/fs_calls.rs]: add a TODO here

File: src/safeposix/syscalls/fs_calls.rs [hunk 5]
```diff
@@ -3058,91 +3144,180 @@ impl Cage {
...
-                                    );
-                                    FS_METADATA.inodetable.remove(&inodenum);
-                                    NET_METADATA.domsock_paths.remove(&path);
+                            // Retrieve the inode object for the socket and decrement its reference
+                            // count. If both the reference count and
+                            // link count are zero, the socket is no longer needed.
+                            let mut inodeobj = FS_METADATA.inodetable.get_mut(&ui.inode).unwrap();
+                            if let Inode::Socket(ref mut sock) = *inodeobj {
+                                sock.refcount -= 1;
+                                if sock.refcount == 0 {
```
Comment (davidge20) [inline src/safeposix/syscalls/fs_calls.rs]: Shouldn't we combine the two if statements with an AND operator?

File: src/safeposix/syscalls/fs_calls.rs [hunk 6]
```diff
@@ -3058,91 +3144,180 @@ impl Cage {
...
                                     .unwrap();
+                                // Link count as 0 represents that there are no hard links present
+                                // for the file, so we need to remove it from the filesystem.
                                 if normalfile_inode_obj.linkcount == 0 {
                                     drop(inodeobj);
-                                    //removing the file from the entire filesystem (interface,
+                                    // removing the file from the entire filesystem (interface,
                                     // metadata, and object table)
                                     FS_METADATA.inodetable.remove(&inodenum);
+                                    // FILEDATAPREFIX represents the common prefix of the name
```
Comment (davidge20) [inline src/safeposix/syscalls/fs_calls.rs:3321]: nice!

File: src/tests/fs_tests.rs [hunk 7]
```diff
@@ -4036,4 +4036,322 @@ pub mod fs_tests {
         assert_eq!(cage.exit_syscall(EXIT_SUCCESS), EXIT_SUCCESS);
         lindrustfinalize();
     }
+
```
Comment (davidge20) [inline src/tests/fs_tests.rs:4069]: great exhaustive testing!

**Review by davidge20 — COMMENTED** [review]
minor comments, overall great job explaining everything in detail!

File: src/safeposix/syscalls/fs_calls.rs (hunk 3 above)
Comment (namanlalitnyu) [reply src/safeposix/syscalls/fs_calls.rs]: Updated!

File: src/safeposix/syscalls/fs_calls.rs (hunk 4 above)
Comment (namanlalitnyu) [reply src/safeposix/syscalls/fs_calls.rs]: Updated!

File: src/safeposix/syscalls/fs_calls.rs (hunk 5 above)
Comment (namanlalitnyu) [reply src/safeposix/syscalls/fs_calls.rs]: Updated.

File: src/safeposix/syscalls/fs_calls.rs [hunk 8]
```diff
@@ -2443,62 +2443,166 @@ impl Cage {
...
+    ///
+    /// The `lseek_syscall()` receives three arguments:
+    /// * `fd` - It is the file descriptor, which identifies the file being
+    ///   accessed
+    /// * `offset` - It determines the position in the file where the file
+    ///   offset should be set. It specifies the number of bytes from a
+    ///   particular reference point.
+    /// * `whence` - It determines how the seek operation is performed. There
+    ///   are three possible values for whence: `SEEK_SET`: The file offset is
+    ///   set to offset bytes. `SEEK_CUR`: The file offset is set to its current
```
Comment (Anway-Agte) [inline src/safeposix/syscalls/fs_calls.rs]: 0 + offset would clarify it even better for the SEEk_SET option

File: src/safeposix/syscalls/fs_calls.rs [hunk 9]
```diff
@@ -3058,52 +3146,126 @@ impl Cage {
...
+    /// For more detailed description of all the commands and return values, see
+    /// [close(2)](https://man7.org/linux/man-pages/man2/close.2.html)
     pub fn close_syscall(&self, fd: i32) -> i32 {
-        //check that the fd is valid
+        // Call the helper function which validates the result and deletes
+        // reference of the file descriptor.
         return Self::_close_helper(self, fd);
     }
 
+    /// ## ----------------- CLOSE HELPER INNER ------------------
```
Comment (Anway-Agte) [inline src/safeposix/syscalls/fs_calls.rs:3183]: is there reason we have a helper function as a separate entity?

File: src/safeposix/syscalls/fs_calls.rs [hunk 10]
```diff
@@ -3058,52 +3146,126 @@ impl Cage {
...
+    /// Upon successful completion of this call, it returns 0. Otherwise, it
+    /// returns error number set with the appropriate error code.
+    ///
+    /// ### Errors
+    ///
+    /// * The syscall returns errors which occur in its helper functions.
+    ///
+    /// ### Panics
+    ///
+    /// * There are no such panics which happen inside the function.
```
Comment (Anway-Agte) [inline src/safeposix/syscalls/fs_calls.rs]: this function indirectly panics when the helper panics - adding that would be good I guess

File: src/safeposix/syscalls/fs_calls.rs [hunk 11]
```diff
@@ -3058,52 +3146,126 @@ impl Cage {
...
+                            // link count are zero, the socket is no longer needed.
+                            let mut inodeobj = FS_METADATA.inodetable.get_mut(&ui.inode).unwrap();
+                            if let Inode::Socket(ref mut sock) = *inodeobj {
+                                sock.refcount -= 1;
+                                if sock.refcount == 0 && sock.linkcount == 0 {
+                                    // Remove the socket from the inode table and the domain
+                                    // socket paths if it is no longer needed.
                                     drop(inodeobj);
                                     let path = normpath(
                                         convpath(sockhandle.localaddr.unwrap().path()),
```
Comment (Anway-Agte) [inline src/safeposix/syscalls/fs_calls.rs:3273]: what do lines 3271 - 3281 do?

File: src/tests/fs_tests.rs [hunk 12]
```diff
@@ -4036,4 +4036,322 @@ pub mod fs_tests {
...
+        // Test to create a file and attempt to seek before the start of the file
+        let fd = cage.open_syscall("/test_file", O_CREAT | O_RDWR, S_IRWXA);
+        assert!(fd >= 0);
+
+        // Attempt to seek to a negative offset and check if it returns an error
+        assert_eq!(
+            cage.lseek_syscall(fd, -10, SEEK_SET),
+            -(Errno::EINVAL as i32)
+        );
```
Comment (Anway-Agte) [inline src/tests/fs_tests.rs:4176]: adding a test case that uses seek_curr with a negative offset that falls before 0 would also be good I guess?

Comment (namanlalitnyu) [reply src/tests/fs_tests.rs:4176]: Added

File: src/safeposix/syscalls/fs_calls.rs (hunk 11 above)
Comment (namanlalitnyu) [reply src/safeposix/syscalls/fs_calls.rs:3273]: Added the comments

File: src/safeposix/syscalls/fs_calls.rs (hunk 10 above)
Comment (namanlalitnyu) [reply src/safeposix/syscalls/fs_calls.rs]: Updated

File: src/safeposix/syscalls/fs_calls.rs (hunk 9 above)
Comment (namanlalitnyu) [reply src/safeposix/syscalls/fs_calls.rs:3183]: Yes, this helper function is called by other syscalls as well. So, that's why we can't club it into one.

File: src/safeposix/syscalls/fs_calls.rs (hunk 8 above)
Comment (namanlalitnyu) [reply src/safeposix/syscalls/fs_calls.rs]: Done

**Comment by namanlalitnyu** [comment]
@davidge20 @Anway-Agte, I have addressed the review comments. Please can you review the changes once?

**Review by davidge20 — APPROVED** [review]
lgtm !

**Comment by namanlalitnyu** [comment]
@yashaswi2000, requesting your review of these changes. 
Thanks!

**Review by Anway-Agte — APPROVED** [review]

**Review by yashaswi2000 — APPROVED** [review]
Great Job, nice comments!

**Comment by namanlalitnyu** [comment]
@rennergade, requesting your review of these changes.
Thanks!

File: src/safeposix/syscalls/fs_calls.rs [hunk 13]
```diff
@@ -3058,91 +3146,181 @@ impl Cage {
...
+            // We decide, how to proceed depending on the fd type.
+            // First we check in the file descriptor to handle stream / epoll (no-op),
+            // sockets (clean the socket), and pipes (clean the pipe), and if it is a
+            // normal file descriptor we decrement the reference count to reflect
             // one less reference to the file.
             match filedesc_enum {
                 //if we are a socket, we dont change disk metadata
-                Stream(_) => {}
-                Epoll(_) => {} //Epoll closing not implemented yet
+                Stream(_) => {} // Stream closing not supported
```
Comment (rennergade) [inline src/safeposix/syscalls/fs_calls.rs]: I think theres some misinterpretation through out here. Its not that closing isnt supported, its just that there is no additional cleanup to do. Of course all file descriptors can be closed.

**Review by rennergade — CHANGES_REQUESTED** [review]
This looks quite good though the comment I pointed out is pretty major. I'd double check everything to be consistent with that.

File: src/safeposix/syscalls/fs_calls.rs (hunk 13 above)
Comment (namanlalitnyu) [reply src/safeposix/syscalls/fs_calls.rs]: Oh, I got it. Thanks for pointing this out, I have updated the comment and checked for other cases as well.

**Review by rennergade — APPROVED** [review]

**Review by rupeshkoushik07 — APPROVED** [review]
